### PR TITLE
Support CQ ID  for any TT-NN Operation

### PR DESCRIPTION
**Pull request by ayerofieiev-tt**
### Ticket
None

### Problem description
Half of TT-NN operations support CQ_ID
Half pretend to support, but don't really support it because don't fully propagate cq_id value in a composite chain.

### What's changed
Introduce 2 new metal apis (**threadlocal**):

```cpp
void PushCurrentCommandQueueId(uint8_t cq_id);
uint8_t PopCurrentCommandQueueId();
uint8_t GetCurrentCommandQueueId();
```
Push and Pop (thanks @dmakoviichuk-tt) allow you to easily change / revert the value. 
In contrast to Get/Set it allows to accomplish things in just 2 calls instead of 3 (`push, pop` instead of `get old value, set new, set old`)

TT-NN api adds another method
```cpp
uint8_t get_current_command_queue_id();
void push_command_queue_id(uint8_t cq_id);
void pop_command_queue_id();
ScopeGuard with_command_queue_id(uint8_t cq_id);
ScopeGuard with_command_queue_id(uint8_t cq_id, lambda);
```

The last lambda allows you things like this
```cpp
ttnn::with_command_queue_id(workload_dispatch_cq, [&](auto guard) {
   output_tensor = ttnn::neg(output_tensor);
});
```

> [!Note]
> I am not a big fan of globals and threadlocals, but I think this is way better than a having a feature that will never work.

Introduced a Context and a Hook in Python level to allow calls like
```py
with ttnn.command_queue(1):
    ttnn.operation1(tensor)           # No cq_id → uses current global cq_id (1)
    ttnn.operation2(tensor, cq_id=0)  # Has cq_id → uses 3, then restores to 0
    ttnn.operation3(tensor)           # No cq_id → uses current global cq_id (1)
```
**Context** allows `with:` calls.

**Hook** captures all calls to a registered TT-NN operation
* handles cq_id kwarg by setting current cq_id
* restores it back after the op finished
* The benefit is that it means ALL ops now support cq_id. 
* **The cost is** that we now do 2 calls from py to cpp instead of 1 when cq_id is specified. 
* If many calls are made it is better to use `with` clause.

> [!Note]
> Things are kind of nice in python here but not in C++. For C++ to be nice too, we need to return a guard. Not sure if guard should be returned in metal api level or should be a TT-NN candy. Please share your thoughts.

> [!Note]
> This PR does not remove handling of CQ ID from ops in C++ yes, and does not update bindings yet. I want to merge this change once we confirm things work as intended. I prefer to make a separate PR for removal of QueueId param from ops and bindings.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/17285081865) CI passes
- [ ] [demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/17285656474)
- [ ] [t3k unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/17285655303)
- [ ] [single card model perf](https://github.com/tenstorrent/tt-metal/actions/runs/17285658396)
- [ ] [t3k model perf](https://github.com/tenstorrent/tt-metal/actions/runs/17285654220)
- [ ] [t3k frequent](https://github.com/tenstorrent/tt-metal/actions/runs/17285681387)

### CODE REVIEW

File: tt_metal/third_party/umd
Comment (omilyutin-tt) [inline tt_metal/third_party/umd:1]: intentional?

Comment (ayerofieiev-tt) [reply tt_metal/third_party/umd:1]: no, thank you

**Comment by ayerofieiev-tt** [comment]
@tt-ai please review

**Comment by ayerofieiev-tt** [comment]
**Claude finished @ayerofieiev-tt's task** —— [View job](https://github.com/tenstorrent/tt-metal/actions/runs/17285069355)

---
### Code Review for CQ ID Support

Completed comprehensive code review for PR #27558 "Support CQ ID for any TT-NN Operation"

#### Analysis Summary:
This PR introduces threadlocal command queue ID management to ensure all TT-NN operations properly support `cq_id` parameters. The implementation adds:

1. **C++ Layer**: Threadlocal storage for current CQ ID with getter/setter functions
2. **Python Layer**: Context manager and hooks to automatically handle `cq_id` parameters
3. **Bindings**: Python bindings for the C++ functions
4. **Hook System**: Pre/post operation hooks to manage CQ ID state automatically

#### Detailed Review Findings:

##### ✅ **Positive Aspects:**

1. **Clean C++ API Design** - The threadlocal approach in `tt_metal.cpp:1426-1431` is simple and thread-safe
2. **Comprehensive Python Integration** - The decorators and context manager provide excellent usability
3. **Backward Compatibility** - Existing code continues to work without changes
4. **Good Documentation** - Python docstrings explain the usage patterns clearly

##### ⚠️ **Areas of Concern:**

**1. Error Handling in Hook System** (`ttnn/ttnn/decorators.py:120, 132`)
- The `cq_id_pre_hook` function logs but doesn't propagate exceptions from `get_current_command_queue_id()` or `set_current_command_queue_id()`
- If these calls fail, the operation continues with potentially incorrect CQ state
- **Recommendation**: Consider failing fast or providing better error recovery

**2. Stack Management Edge Cases** (`ttnn/ttnn/decorators.py:142-150`)  
- The `CQ_ID_STACK` could become inconsistent if operations throw exceptions between pre/post hooks
- No validation that stack state matches expectations in `cq_id_post_hook`
- **Recommendation**: Add stack validation and consider using try/finally patterns

**3. Context Manager Robustness** (`ttnn/ttnn/decorators.py:176-183`)
- Warning message suggests potential issues with operations not properly restoring CQ state
- The detection logic might not catch all edge cases where CQ ID changes unexpectedly
- **Recommendation**: Consider more robust state validation or clearer failure modes

**4. Performance Considerations**
- Every TT-NN operation now has 2 additional hook calls when `cq_id` is specified
- Hook system is enabled globally by default (`decorators.py:191-192`)
- **Recommendation**: Consider performance impact measurement and potential opt-out mechanisms for performance-critical code

**5. Global State Management**
- Heavy reliance on global state (`CQ_ID_STACK`, `PRE_OPERATION_HOOKS`, `POST_OPERATION_HOOKS`)
- Could complicate debugging and testing scenarios
- **Recommendation**: Consider encapsulating state in a context object

##### 🔍 **Code-Specific Issues:**

**Bug in `ttnn/ttnn/decorators.py:120`:**
```python
logger.info(f"Using cq_id {function_kwargs['cq_id']} from operation kwargs")
```
This line references `function_kwargs['cq_id']` after it was already popped on line 119, causing a KeyError.

**Suggested fix:**
```python
cq_id = function_kwargs.pop("cq_id")
logger.info(f"Using cq_id {cq_id} from operation kwargs")
```

**Logic Issue in `ttnn/ttnn/decorators.py:142-144`:**
```python
prev_cq_id = CQ_ID_STACK[-1]
if prev_cq_id is None:
    return None
```
The stack contains the old CQ ID, not None. This check should verify the stack isn't empty instead.

##### 📝 **Minor Issues:**

1. **Documentation**: The C++ functions in `host_api.hpp:967-976` could benefit from more detailed documentation about thread safety and typical usage patterns
2. **Naming**: Consider whether "current" command queue implies thread-local vs global scope in the API names
3. **Type Safety**: The `uint8_t cq_id` parameter could benefit from bounds checking or typedef for clearer intent

##### 🚀 **Recommendations:**

1. **Fix the KeyError bug** in the logging statement
2. **Add exception handling** around critical CQ ID operations in hooks
3. **Consider performance testing** with the new hook overhead
4. **Add unit tests** for edge cases in stack management and context managers
5. **Document thread safety guarantees** more explicitly in C++ headers

##### 📊 **Overall Assessment:**

This is a well-architected solution to a complex problem. The approach of using threadlocal storage with Python hooks provides good usability while maintaining backward compatibility. The main concerns are around error handling robustness and the performance impact of the global hook system.

The PR successfully addresses the core issue where TT-NN operations had inconsistent CQ ID support. The Python API design is intuitive and should significantly improve developer experience.

**Recommendation: Approve with minor fixes** - particularly the KeyError bug and improved error handling in the hook system.

**Review by TT-BrianLiu — APPROVED** [review]

File: ttnn/ttnn/decorators.py [hunk 1]
```diff
@@ -98,6 +101,99 @@ def register_pre_operation_hook(hook):
...
 POST_OPERATION_HOOKS = []
 
+# Stack to track cq_id contexts and restoration info
+# Below hook and context manager allows user to do:
+# with ttnn.command_queue(5):
+#    ttnn.operation1(tensor)           # No cq_id → uses current global cq_id (5)
+#    ttnn.operation2(tensor, cq_id=3)  # Has cq_id → uses 3, then restores to 5
+#    ttnn.operation3(tensor)           # No cq_id → uses current global cq_id (5)
+
+CQ_ID_STACK = []
```
Comment (dmakoviichuk-tt) [inline ttnn/ttnn/decorators.py:110]: That what I was thinking, you should jhave in c++.
Instead of set you can do old OpenGL style  with push/pop attributes.

File: tt_metal/api/tt-metalium/host_api.hpp [hunk 2]
```diff
@@ -955,6 +955,25 @@ bool EventQuery(const std::shared_ptr<Event>& event);
 void Synchronize(
     IDevice* device, std::optional<uint8_t> cq_id = std::nullopt, tt::stl::Span<const SubDeviceId> sub_device_ids = {});
 
+// clang-format off
```
Comment (omilyutin-tt) [inline tt_metal/api/tt-metalium/host_api.hpp:958]: I'd really prefer another header where this feature is better documented.

Also core pybind can bind that function directly, we don't need an extra hop?

**Review by omilyutin-tt — APPROVED** [review]
Also correct PR description:

with ttnn.command_queue(1):
    ttnn.operation1(tensor)           # No cq_id → uses current global cq_id (5)

File: tt_metal/api/tt-metalium/device.hpp [hunk 3]
```diff
@@ -148,7 +149,9 @@ class IDevice {
     virtual uint32_t get_noc_multicast_encoding(uint8_t noc_index, const CoreRange& cores) const = 0;
 
     virtual SystemMemoryManager& sysmem_manager() = 0;
-    virtual CommandQueue& command_queue(size_t cq_id = 0) = 0;
+
+    // If cq_id is not provided, the current command queue is returned from the current thread
+    virtual CommandQueue& command_queue(std::optional<uint8_t> cq_id = std::nullopt) = 0;
```
Comment (dmakoviichuk-tt) [inline tt_metal/api/tt-metalium/device.hpp:154]: what worries me. Current command_queue() call means command_queue(0) in your case it means another. Are you sure fixed everything?

File: ttnn/ttnn/decorators.py (hunk 1 above)
Comment (ayerofieiev-tt) [reply ttnn/ttnn/decorators.py:110]: In C++ I wanted to go with simple Guards and wanted to keep it out of scope of this PR

File: tt_metal/tt_metal.cpp [hunk 4]
```diff
@@ -1422,6 +1422,14 @@ void Synchronize(IDevice* device, const std::optional<uint8_t> cq_id, tt::stl::S
     }
 }
 
+namespace {
+thread_local uint8_t current_command_queue_id = 0;
```
Comment (dmakoviichuk-tt) [inline tt_metal/tt_metal.cpp:1426]: are you sure you want to make it thread_local? different thread might have different cq_id? so if I call it from main thread I can get one if I call from some worker thread I get another value?

File: tt_metal/api/tt-metalium/device.hpp (hunk 3 above)
Comment (ayerofieiev-tt) [reply tt_metal/api/tt-metalium/device.hpp:154]: yes. it still means 0 unless someone changed the current

File: tt_metal/tt_metal.cpp (hunk 4 above)
Comment (ayerofieiev-tt) [reply tt_metal/tt_metal.cpp:1426]: yes. i think it makes sense from expectations standpoint.
besides that it makes things easier as it requires no sync

File: tt_metal/api/tt-metalium/host_api.hpp (hunk 2 above)
Comment (ayerofieiev-tt) [reply tt_metal/api/tt-metalium/host_api.hpp:958]: It would be really great for Metal to have its own bindings
But so far there are no bindings on metal level and I don't want to introduce any in this PR.

File: tt_metal/tt_metal.cpp (hunk 4 above)
Comment (dmakoviichuk-tt) [reply tt_metal/tt_metal.cpp:1426]: Id make it atomic instead.

**Review by dmakoviichuk-tt — APPROVED** [review]
I provided a few more ideas in a call.
And a last one:
Please create Context class/struct and put your cq_id inside. It is hard to search for an alone global vars
